tests/tape/interfaces/test_qnode_torch.py

```diff
@@ -527,6 +528,149 @@ def circuit():
+            0.5 - 0.5 * np.cos(a) * np.cos(b)
+        ]
+        assert np.allclose(res.detach(), expected_res, atol=tol, rtol=0)
+
+        expected_g = [
+            [-0.5 * np.sin(a) * np.cos(b), -0.5 * np.cos(a) * np.sin(b)],
+            [0.5 * np.sin(a) * np.cos(b), 0.5 * np.cos(a) * np.sin(b)]
+        ]
+        assert np.allclose(g.detach(), expected_g, atol=tol, rtol=0)
+
+        expected_hess = [
+            [
+                [-0.5 * np.cos(a) * np.cos(b), 0.5 * np.sin(a) * np.sin(b)],
+                [0.5 * np.sin(a) * np.sin(b), -0.5 * np.cos(a) * np.cos(b)]
+            ],
+            [
+                [0.5 * np.cos(a) * np.cos(b), -0.5 * np.sin(a) * np.sin(b)],
+                [-0.5 * np.sin(a) * np.sin(b), 0.5 * np.cos(a) * np.cos(b)]
+            ]
+        ]
+        assert np.allclose(hess.detach(), expected_hess, atol=tol, rtol=0)
+
+    def test_hessian_ragged(self, dev_name, diff_method, mocker, tol):
+        """Test hessian calculation of a ragged QNode"""
+        if diff_method not in {"parameter-shift", "backprop"}:
+            pytest.skip("Test only supports parameter-shift or backprop")
+
+        dev = qml.device(dev_name, wires=2)
+
+        @qnode(dev, diff_method=diff_method, interface="torch")
+        def circuit(x):
+            qml.RY(x[0], wires=0)
+            qml.RX(x[1], wires=0)
+            qml.RY(x[0], wires=1)
+            qml.RX(x[1], wires=1)
+            return qml.expval(qml.PauliZ(0)), qml.probs(wires=1)
+
+        x = torch.tensor([1.0, 2.0], requires_grad=True)
+        res = circuit(x)
+        jac_fn = lambda x: jacobian(circuit, x, create_graph=True)
+
+        g = jac_fn(x)
+
+        spy = mocker.spy(JacobianTape, "hessian")
+        hess = jacobian(jac_fn, x)
+        spy.assert_called_once()
+
+        a, b = x.detach().numpy()
+
+        expected_res = [
+            np.cos(a) * np.cos(b),
+            0.5 + 0.5 * np.cos(a) * np.cos(b),
+            0.5 - 0.5 * np.cos(a) * np.cos(b)
+        ]
+        assert np.allclose(res.detach(), expected_res, atol=tol, rtol=0)
+
+        expected_g = [
+            [-np.sin(a) * np.cos(b), -np.cos(a) * np.sin(b)],
+            [-0.5 * np.sin(a) * np.cos(b), -0.5 * np.cos(a) * np.sin(b)],
+            [0.5 * np.sin(a) * np.cos(b), 0.5 * np.cos(a) * np.sin(b)]
+        ]
+        assert np.allclose(g.detach(), expected_g, atol=tol, rtol=0)
+
+        expected_hess = [
+            [
+                [-np.cos(a) * np.cos(b), np.sin(a) * np.sin(b)],
+                [np.sin(a) * np.sin(b), -np.cos(a) * np.cos(b)]
+            ],
+            [
+                [-0.5 * np.cos(a) * np.cos(b), 0.5 * np.sin(a) * np.sin(b)],
+                [0.5 * np.sin(a) * np.sin(b), -0.5 * np.cos(a) * np.cos(b)]
+            ],
+            [
+                [0.5 * np.cos(a) * np.cos(b), -0.5 * np.sin(a) * np.sin(b)],
+                [-0.5 * np.sin(a) * np.sin(b), 0.5 * np.cos(a) * np.cos(b)]
+            ]
+        ]

```

Comment:
So the qnode is ragged, but the Hessian is a more standard `(3, 2, 2)` shape? Is this what PyTorch would expect?
E.g., not something like `[hessian1, hessian2]` with `hessian1` of shape `(2, 2)` and `hessian2` of shape `(2, 2, 2)`?

I don't mind either way, just wondering.

Reply:
This is a very good point.

Currently, if the QNode returns 'ragged' output (e.g., expval + probs, or multiple probs on different wires), we are flattening the output due to a very ancient restriction we had that QNodes _always_ return NumPy arrays. This restriction was made back when we only had an Autograd interface to support, and no `probs`.

However, it's probably time we updated this:

- If the QNode `return` statement returns a tuple, the user should _get a tuple_ as output, not a NumPy array.
- This avoids the whole issue of ragged arrays altogether

Reply:
This would actually be a cool thing to sort out now-ish, before usage of `return expval, probs` etc. becomes entrenched (at the moment, it's quite rare)

Reply:
I got curious, and coded this up.

## TensorFlow

```python
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev, interface="tf")
def circuit(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0)), qml.probs(wires=[0, 1])

x = tf.Variable(0.5, dtype=tf.float64)

with tf.GradientTape() as tape:
    res = circuit(x)
```

where:

```
>>> print(res)
(<tf.Tensor: shape=(), dtype=float64, numpy=0.8775825618903726>, <tf.Tensor: shape=(4,), dtype=float64, numpy=array([0.93879128, 0.        , 0.06120872, 0.        ])>)
```

Unfortunately, TF does not let you call Jacobian on a tuple :(
```python
Traceback (most recent call last):
  File "test.py", line 54, in <module>
    grad = tape.jacobian(res, x)
  File "/home/josh/miniconda3/envs/37/lib/python3.7/site-packages/tensorflow/python/eager/backprop.py", line 1166, in jacobian
    target_static_shape = target.shape
AttributeError: 'tuple' object has no attribute 'shape'
```

You can do `grad = [tape.jacobian(r, x) for r in res]`, but it is slower and you lose vectorization.


## Torch

I think this is fully possible, but can't get the backwards call working. Sounds like it would result in a much more complicate torch `backward(grad_output)` method, since we can no longer assume that `grad_output` is flat - it will now have to match the 'shape' of the forward pass output tuple.

Reply:
Thanks @josh146, interesting investigation! Looks like returning a tuple is more expected, but harder to support with the different interfaces. In which case, I'm happy to keep it as is here and maybe we come back to think about this question.

Do we expect this use case to become common, e.g., users regularly requesting `return expval, probs`? In practice, I've typically requested `probs` in isolation.

Reply:
I think returning a tuple can be supported on all interfaces, it just requires a bit more work. But definitely doable.

> I've typically requested probs in isolation.

agree.

There is one case I am worried about though:

```python
return expval(..), expval(..)
```

Currently this returns a length-2 numpy array. Under the new suggestion above, this would return a tuple of numpy scalars. I think this is more intuitive from a UI perspective, but would be a _very_ invasive change.

Reply:
> There is one case I am worried about though

Good point, although I'd worry that making such a change may be annoying since probably most use cases involve treating the output as a numpy array even if it was technically specified as a tuple return.

Reply:
> I'd worry that making such a change may be annoying

It might not be... indexing `res[0]` will continue to work with a tuple, and NumPy functions such as `np.sum(res)` also work with tuples. It might... be okay?

Definitely having the QNode always return a tuple will be better for long term flexibility